### PR TITLE
Improve peer indexing

### PR DESCRIPTION
**Pull request by jcorbin**
This sets the stage to add more indices towards #66:
- factor out general `{add,delete}IndexEntry` utilities
- factor out `{add,delete}PeerIndex` methods
- delete index entries in `reapSinglePeer` to protect against double reap edge cases

r @kriskowal @Raynos

### CODE REVIEW

**Comment by jcorbin** [comment]
Interesting, so the commit that breaks test is https://github.com/uber/hyperbahn/commit/5490c36e8c0df1ae72cf5bd7fdce6d5b1d169a57, which I think is more correct behavior, but I need to investigate more as to why `test/forward/dead-remote-reaped.js` fails with it.

File: service-proxy.js [hunk 1]
```diff
@@ -1149,3 +1155,31 @@ function extendLogInfo(info) {
...
+}
+
+/* eslint-disable guard-for-in, no-unused-vars */
+function isObjectEmpty(obj) {
+    for (var prop in obj) {
+        return false;
+    }
+    return true;
+}
+/* eslint-enable guard-for-in, no-unused-vars */
```
Comment (Raynos) [inline service-proxy.js]: Soon line 1200. SOON. Then 1500, then 2250, then finally 6000.

Comment (jcorbin) [inline service-proxy.js]: Couldn't agree more... this whole file needs to get re-orged and has grown too much... eventually...

**Comment by Raynos** [comment]
lgtm.
